### PR TITLE
tidb, variable: refactor retry, add TransactionContext

### PR DESCRIPTION
**Pull request by coocood**
Add TransactionContext which is used to store transaction scope variables, so we can clean up multiple variables at once.

Refactor session Retry related code, improves readability.

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -239,45 +221,47 @@ func (s *session) finishTxn(rollback bool) error {
 			s.txn.SetOption(kv.BinlogData, bin)
 		}
 	}
-
 	if err := s.checkSchemaValid(); err != nil {
-		if !s.sessionVars.RetryInfo.Retrying && s.isRetryableError(err) {
```
Comment (tiancaiamao) [inline session.go]: oh, I see... it's just moved to somewhere else...

File: session_test.go [hunk 2]
```diff
@@ -2381,41 +2379,6 @@ func (s *testSessionSuite) TestSqlLogicTestCase(c *C) {
...
-		values:      make(map[fmt.Stringer]interface{}),
-		store:       store,
-		debugInfos:  make(map[string]interface{}),
-		maxRetryCnt: 10,
-		sessionVars: variable.NewSessionVars(),
-	}
-	return s
-}
-
-func (s *testSessionSuite) TestRetryAttempts(c *C) {
```
Comment (coocood) [inline session_test.go]: yes

**Comment by tiancaiamao** [comment]
LGTM

File: session.go [hunk 3]
```diff
@@ -594,6 +564,7 @@ func (s *session) ExecutePreparedStmt(stmtID uint32, args ...interface{}) (ast.R
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	PrepareTxnCtx(s)
 	st := executor.CompileExecutePreparedStmt(s, stmtID, args...)
```
Comment (zimulala) [inline session.go]: `executor.GetInfoSchema(s)` can be used in the function of `CompileExecutePreparedStmt`.

File: session.go [hunk 4]
```diff
@@ -209,22 +199,14 @@ func (s *session) SetConnectionID(connectionID uint64) {
...
-func (s *session) finishTxn(rollback bool) error {
-	// transaction has already been committed or rolled back
+func (s *session) doCommit() error {
 	if s.txn == nil {
 		return nil
 	}
 	defer func() {
 		s.txn = nil
 		s.sessionVars.SetStatusFlag(mysql.ServerStatusInTrans, false)
-		binloginfo.ClearBinlog(s)
```
Comment (zimulala) [inline session.go]: Is it safe to remove clear binlog value?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**